wsgi dashboards plugin

tardyp opened the pull request.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,72 @@ For more information on the available commands, see :doc:`../developer/master-wo
 
-Writing New Status Plugins
---------------------------
+Writing Dashboards with Flask
```
Interesting about Flask instead of something like Klein. I do get why; Flask is vastly more popular and better understood.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,72 @@ For more information on the available commands, see :doc:`../developer/master-wo
+
+.. code-block:: python
 
-Each status plugin is an object which provides the :class:`twisted.application.service.IService` interface, which creates a tree of Services with the buildmaster at the top [not strictly true].
```
Is this text outdated or present somewhere else? I don't see it being moved elsewhere.

tardyp replied:
yes, its talking about status services

Frodox replied:
Yup. *How to write custom plugin* still needed :)

Frodox commented:
Great +1. It is just about me. I want to a bit restructure/customize build results page, but I haven't time now to learn few frameworks.. Want to test this plugin very much.
How powerful is it..?

Frodox commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,72 @@ For more information on the available commands, see :doc:`../developer/master-wo
+    c['www']['plugins']['flask_dashboard'] = {
+        'dashboards': [
+            {
+                'name': 'mydashboard',  # as used in URLs
```
Can I use several dashboards, identified by `name` ..? May be `templates/name.html` for this case..?

tardyp replied:
the name of the template is defined by the app endpoint

Frodox commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,72 @@ For more information on the available commands, see :doc:`../developer/master-wo
+
+Then you need a ``templates/index.html`` file near your ``master.cfg``
+
+.. code-block:: html
```
It is flask template language, or something else..?

tardyp replied:
flask uses jinja as its templating langage

tardyp commented:
- Updated the documentation to give more info
- Renamed the plugin as it is not bound to flask (any wsgi can work)
- added support for buildsummary and buildsticker directives

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
+
+Buildbot Nine UI is written in Javascript.
+This allows it to be reactive and real time, but comes at a price of a fair complexity.
+Sometimes, you just need a dashboard displaying your build results in your own weird manner but learning AngularJS for that is just too much.
```
Remove "weird" and "just". Using "just" can makes things seem more trivial than the reader thinks them. It is frequently better to omit it. Using "weird" can seem overly judging. They don't think it's weird and instead may think it weird that it's so hard to do what they consider normal.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
+  No need to worry about Twisted and asynchronous code.
+  You can use python-requests_ or any library from the python ecosystem to access other servers.
+
+- You could use Http in order to access Buildbot :ref:`REST_API`, but you can also use the :ref:`Data_API`, via the provided synchronous wrapper.
```
Make Http all caps.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
+
+- That html code output of the server runs inside AngularJS application.
+
+  - It will use the CSS of the AngularJS application (including the Bootstrap_ css base).
```
Make "css" all caps to be consistent within the line.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
+- That html code output of the server runs inside AngularJS application.
+
+  - It will use the CSS of the AngularJS application (including the Bootstrap_ css base).
+  - It can use some of the AngularJS directives defined by buildbot UI (currently only buildsummary is usable).
```
Buildbot instead of buildbot.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
+  - It can use some of the AngularJS directives defined by buildbot UI (currently only buildsummary is usable).
+  - It has full access to the application JS context.
+
+Here is an example of code that you can use in your master.cg to create a simple dashboard:
```
master.cfg instead of master.cg.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
+
+Here is an example of code that you can use in your master.cg to create a simple dashboard:
+
+.. code-block:: python
```
Has this full block been confirmed to work? Would it be possible to add a test for it?

tardyp replied:
I tested it manually. It needs a test in e2e smokes in order to really test it.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
+        }
+    ]
+
+Then you need a ``templates/mydashboard.html`` file near your ``master.cfg``
```
End sentence with a period.

seankelly commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,153 @@ For more information on the available commands, see :doc:`../developer/master-wo
-The status plugins are all children of an object which implements :class:`buildbot.interfaces.IStatus`, the main status object.
-From this object, the plugin can retrieve anything it wants about current and past builds.
-It can also subscribe to hear about new and upcoming builds.
+Note that those two directives will make additional http requests from the browser in order to fetch the necessary data they need to be rendered.
```
s/http/HTTP/.

seankelly commented:
I really like this. I think it will help adoption a lot.

tardyp commented:
fixes according to review. thanks @seankelly !

- rephrasing
- e2e test for wsgi_dashboard

tardyp commented on smokes/run_docker.sh:
```diff
@@ -1,32 +0,0 @@
-#!/bin/bash
```
this would not work with the templates directory, and was not used anyway.

Frodox commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,158 @@ For more information on the available commands, see :doc:`../developer/master-wo
+
+- The application framework runs in a thread outside of Twisted.
+  No need to worry about Twisted and asynchronous code.
+  You can use python-requests_ or any library from the python ecosystem to access other servers.
```
Could you add example, please..?

tardyp replied:
I added an example with requests usage, and a line chart. I hope the example does not get over complicated because of that. please let me know.

Frodox replied:
No, it is clear enough.  Thank  you, looks great 👍

Frodox commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,83 @@ For more information on the available commands, see :doc:`../developer/master-wo
+
+- That html code output of the server runs inside AngularJS application.
+
+  - It will use the CSS of the AngularJS application (including the Bootstrap_ CSS base).
```
Is it possible  (how?) to override/extend/use custom css?

Frodox commented on master/docs/manual/customization.rst:
```diff
@@ -932,18 +932,83 @@ For more information on the available commands, see :doc:`../developer/master-wo
+  - It can use some of the AngularJS directives defined by Buildbot UI (currently only buildsummary is usable).
+  - It has full access to the application JS context.
+
+Here is an example of code that you can use in your master.cfg to create a simple dashboard:
```
Backtricks is missed ````

Frodox commented on smokes/master.cfg:
```diff
@@ -115,5 +113,6 @@ c['buildbotNetUsageData'] = None
-    'db_url' : "sqlite:///state.sqlite",
+    'db_url': "sqlite:///state.sqlite",
 }
+exec(open("mydashboard.py").read())
```
Is it original `master.cfg` ? One new line with comments before this exec would  be helpfull

Frodox replied:
Can I use same things in my master.cfg? :)
